src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+                posHold.sanityCheckDistance = gpsSol.groundSpeed > 1000 ? gpsSol.groundSpeed : 1000.0f;
+                // 1s of flight at current speed or 10m, in cm
+            }
+            // primarily to detect flyaway from no Mag or badly oriented Mag
+            // but must accept some overshoot at the start, especially if entering at high speed
+            if (distanceCm > posHold.sanityCheckDistance) {
+                return false; // must stay within 10m or probably flying away
+                // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+                // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+                // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+            }
+    
+            vectors_t *vectors[] = { &posHold.NS, &posHold.EW };
+            for (int i = 0; i < 2; i++) {
+                vectors_t *latLong = vectors[i];
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+    
+                // ** P **
+                float pidP = latLong->distance * positionPidCoeffs.Kp;
+    
+                // ** I **
+                if (!latLong->isStarting){
+                    // only accumulate iTerm after completing the start phase
+                    // perhaps need a timeout on the start phase ?
+                    latLong->integral += latLong->distance * posHold.gpsDataIntervalS;
+                } else {
+                    // while moving sticks, slowly leak iTerm away, approx 2s time constant
+                    latLong->integral *= leak;
+                }
+                float pidI = latLong->integral * positionPidCoeffs.Ki;
+        
+                // ** D ** //
+                // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+                // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
+                float velocity = (latLong->distance - latLong->previousDistance) * posHold.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
+                latLong->previousDistance = latLong->distance;
+    
+                float acceleration = (velocity - latLong->previousVelocity) * posHold.gpsDataFreqHz;
+                latLong->previousVelocity = velocity;
+        
+                // scale and filter - filter cutoffs vary during the startup phase
+                float pidD = velocity * positionPidCoeffs.Kd;
+                pt1FilterUpdateCutoff(&latLong->velocityLpf, posHold.pt1Gain);
+                pidD = pt1FilterApply(&latLong->velocityLpf, pidD);
+    
+                float pidA = acceleration * positionPidCoeffs.Kd;
+                pt2FilterUpdateCutoff(&latLong->accelerationLpf, posHold.pt1Gain);
+                pidA = pt2FilterApply(&latLong->accelerationLpf, pidA);
```

Comment:
Using filtered velocity for acceleration will provide first PT1 state implicitly (and will avoid overusing  pt1 alpha for pt2; but it will attenuate 6dB at cutoff)

Reply:
OK, done, sorry to be slow implementing this